src/source/tile_id.js

```diff
@@ -164,11 +164,11 @@ export class OverscaledTileID {
     }
 }
 
-function calculateKey(wrap: number, z: number, x: number, y: number) {
+function calculateKey(wrap: number, overscaledZ: number, z: number, x: number, y: number) {
     wrap *= 2;
     if (wrap < 0) wrap = wrap * -1 - 1;
     const dim = 1 << z;
-    return ((dim * dim * wrap + dim * y + x) * 32) + z;
+    return (((dim * dim * wrap + dim * y + x) * 32) + z) * 32 + overscaledZ;
```

Comment:
I'm trying to figure out whether we have enough precision in javascript to safely represent this key. 2^53 - 1 is the maximum safe integer js numbers can hold. Each of the zoom components takes 5 bits which leaves 43 bits for the rest. Ignoring the world wrapping part would leave 21 bits for x and y.

It would be rare for this to be a problem but maybe we should switch to strings here to be safe. I'm not sure if there are perf implications.

Or at least `assert(z < 22)`.

Reply:
> Or at least `assert(z < 22)`.

We can't — the default `maxZoom` is `22` but can be bumped further through `options` (e.g. to `24`).

We could try winning a few bytes by encoding the difference between `overscaledZ` and `z` rather than the full `overscaledZ`, which should have a smaller range of possible values.